### ISSUE TITLE
mamba segfault when uninstalling package installed with older mamba version

### ISSUE THREAD

**Issue by wolfv** [mamba-org/mamba#1402]
@consideratio reported the following issue:

I've created a reproducible snippet. I'm going to try seeing if I can reproduce it also if starting out with mamba 0.20 from the start and not just updating to it along the way.

```Dockerfile
# 166a330 is a tag from 2022 jan 19 where a lot of packages were installed with
# mamba 0.17
FROM pangeo/pangeo-notebook:166a330

RUN echo "WILL FAIL: Attempting to remove nomkl with mamba version: $(mamba --version)" \
 && mamba uninstall -n ${CONDA_ENV} -vv -y nomkl || true

RUN mamba --version \
 && mamba update -y mamba \
 && mamba --version

RUN echo "WILL FAIL: Attempting to remove nomkl with mamba version: $(mamba --version)" \
 && mamba uninstall -n ${CONDA_ENV} -vv -y nomkl || true

RUN echo "WILL SUCCEED: Attempting to remove nomkl with mamba version: $(mamba --version)" \
 && echo "BY clearing the history file..." \
 && rm /srv/conda/envs/${CONDA_ENV}/conda-meta/history \
 && mamba uninstall -n ${CONDA_ENV} -vv -y nomkl
```

And

I've now reproduced this when mamba is installed to 0.20 from the beginning and not upgraded to before trying the mamba uninstall -n ${CONDA_ENV} nomkl command which is failing.

@wolfv let me know if you have ideas on how I could contribute with additional information etc.

**Comment by wolfv** [comment]
#1347

**Comment by consideRatio** [comment]
@wolfv while initially I suspected it could be that a lot of packages were installed with old mama, I put in work to investigate if it would happen also if I used mamba 0.20 from the start.

So, I've verified that this was an issue if:
1. miniforge was installed -> a mamba installation followed directly
2. a lot of packages were installed, including `nomkl`
3. I attempted to uninstall `nomkl`

The segfault error showed up right after a statement about parsing history:

```
DEBUG conda.core.prefix_data:_load_single_record(172): loading prefix record /srv/conda/envs/notebook/conda-meta/xcape-0.1.4-py39h34f0aaa_1.json
DEBUG conda.core.prefix_data:_load_single_record(172): loading prefix record /srv/conda/envs/notebook/conda-meta/urllib3-1.26.8-pyhd8ed1ab_1.json
DEBUG conda.core.prefix_data:_load_single_record(172): loading prefix record /srv/conda/envs/notebook/conda-meta/libblas-3.9.0-12_linux64_openblas.json
DEBUG conda.core.prefix_data:_load_single_record(172): loading prefix record /srv/conda/envs/notebook/conda-meta/traitlets-5.1.1-pyhd8ed1ab_0.json
DEBUG conda.core.prefix_data:_load_single_record(172): loading prefix record /srv/conda/envs/notebook/conda-meta/fastapi-0.72.0-pyhd8ed1ab_0.json
info     Reading cache files '/tmp/tmplulgsj8_.*' for repo index 'installed'
info     parsing history: "/srv/conda/envs/notebook/conda-meta/history"
Segmentation fault (core dumped)
```

The error went away by first deleting my history file and then trying to uninstall the `nomkl` package.

**Comment by consideRatio** [comment]
The reproduction can be something like...

```
git clone https://github.com/pangeo-data/pangeo-docker-images
cd pangeo-docker-images

# Modify the base-image/Dockerfile with
# CONDA_VERSION=4.11.0-0
# MAMBA_VERSION=0.20

# Then run from the git repo's root folder
make base-image
make pangeo-notebook
```

Finally, create a new Dockerfile:

```Dockerfile
FROM pangeo/pangeo-notebook:master

RUN echo "WILL FAIL: Attempting to remove nomkl with mamba version: $(mamba --version)" \
 && mamba uninstall -n ${CONDA_ENV} -vv -y nomkl || true

RUN echo "WILL SUCCEED: Attempting to remove nomkl with mamba version: $(mamba --version)" \
 && echo "BY clearing the history file..." \
 && rm /srv/conda/envs/${CONDA_ENV}/conda-meta/history \
 && mamba uninstall -n ${CONDA_ENV} -vv -y nomkl
```

**Comment by wolfv** [comment]
Thanks, perfect reproducers! I'll have a look asap. It might well be that there is still a problem in the history parsing :)

**Comment by jonashaag** [comment]
Minimal reproducer

```
FROM ubuntu:21.04

RUN apt update \
 && apt install -y --no-install-recommends curl gdb

# Maybe change aarch64 to x86_64
RUN curl -vLk https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-aarch64.sh > install-mamba

RUN chmod +x install-mamba \
 && bash ./install-mamba -b \
 && /root/mambaforge/bin/mamba init

RUN /root/mambaforge/bin/mamba create -n x \
 && /root/mambaforge/bin/python -c 'print("==>x<==\n#" + 30000*"x")' > /root/mambaforge/envs/x/conda-meta/history

CMD ["bash", "-c", "gdb --args /root/mambaforge/bin/python /root/mambaforge/bin/mamba uninstall -n x x -v"]
```

**Comment by jonashaag** [comment]
Also happens with ubuntu:20.04, 21.10, but not on OSX.

**Comment by wolfv** [comment]
I am looking into this now

**Comment by wolfv** [comment]
Very strange. I think the segfault is in the `parse_comment_line` function and relates to one of the regex's there. At least for the reproducer of @jonashaag.

**Comment by jonashaag** [comment]
Could there be a problem with `std::regex` on large strings?

**Comment by wolfv** [comment]
well, there shouldn't be but I was able to reproduce this with this regex and (very) long strings: static std::regex spec_pat("#\\s*(\\w+)\\s*specs:\\s*(.+)?");


I'll switch to `find_first_of` and some simpler string processing instead and hopefully that fixes it. @consideRatio do you know by any chance if there is a really long line in your history file?

**Comment by jonashaag** [comment]
@wolfv Yes there is. It's the same root cause that I've found in #1347. In my comment above I simplified the history file to a bunch of `x`s

**Comment by wolfv** [comment]
I pushed a potential fix here: https://github.com/mamba-org/mamba/pull/1441

**Comment by wolfv** [comment]
Any case one of you could try with the latest mamba / micromamba to see if it is still an issue?

**Comment by consideRatio** [comment]
@wolfv nice work! The issue is resolved as of version 0.21.0!

```shell
docker run -it --rm pangeo/pangeo-notebook:master bash

# from within the container...

# updating from mamba 0.20.0 to 0.21.0+ makes us not get segfault
source /srv/conda/bin/activate
mamba install -y mamba=0.21.0
conda deactivate
mamba --version

# causes a segfault with mamba 0.20.0 but not with 0.21.0, 0.21.1, 0.21.2
mamba uninstall -n ${CONDA_ENV} -vv -y nomkl
```

**Comment by wolfv** [comment]
great, thanks for confirming!
